### DISCUSSION TITLE
This design is too complex

### DISCUSSION THREAD

**Discussion by cnwzhu** [tokio-rs/toasty#42] [Ideas]
This design is too complex and requires learning a special DSL. Can it be designed referencing MyBatis in Java?

**Comment by PeterVajda87** [comment]
Would you maybe elaborate on what you consider too complex? I am curious, because I worked with SeaOrm previously and I find this design a magnitude less complex.

**Reply by cnwzhu** [reply]
I hope to design a query method similar to MyBatis/Ibatis, adding support for automatic encapsulation of associated query objects. This approach is very easy to use and does not require specialized learning of the framework.

```
#[html_sql(
r#"<select id="select_by_condition">
        `select * from activity`
        <where>
         <if test="a">
                ` and name like #{name}`
            </if>
            <if test="name != ''">
                ` and name like #{name}`
            </if>
            <if test="dt >= '2023-11-03T21:13:09.9357266+08:00'">
                ` and create_time < #{dt}`
            </if>
            <choose>
                <when test="true">
                    ` and id != '-1'`
                </when>
                <otherwise>and id != -2</otherwise>
            </choose>
            ` and `
            <trim prefixOverrides=" and">
                ` and name != '' `
            </trim>
        </where>
  </select>"#
)]
async fn select_by_condition(
    name: &str,
    dt: &DateTime,
    a: bool,
) -> rbatis::Result<Vec<Activity>> {
    impled!()
}
```

**Reply by cnwzhu** [reply]
With this approach, users only need to focus on their business logic and do not have to worry about the various query techniques of the Rust language.

**Reply by bsodmike** [reply]
Since this is a Rust ORM, I'm in favour of keeping the DSL in Rust.  I'm not sure why a "Rust" ORM should be language agnostic? Please explain?

As far as ORMs go, I've just played with the hello-toasty example, and it's grrrrrreat(TM)!

**Reply by cnwzhu** [reply]
Many developers prefer using SQL because it is intuitive and powerful. If an ORM framework can provide better support for SQL, especially by simplifying result mapping and allowing developers to use SQL directly for queries, it can indeed reduce the learning curve and improve development efficiency.

If a DSL (domain-specific language) design is too complex, it may be confusing, especially in a language like Rust, where the type system and borrow checker already have a certain learning curve. Providing a well-designed SQL interface, combined with type-safe result mapping, could be more popular. This is also an important factor for many developers when choosing an ORM.

**Reply by cnwzhu** [reply]
The learning curve for Rust is already quite steep, and adding a complex ORM DSL may leave developers feeling overwhelmed. Rust’s ownership model, lifetimes, and concurrency model all take time to understand, and adding the extra burden of learning a framework can indeed impact the developer experience.

To improve usability and better promote Rust in the web development field, the following points might be considered:

	1.	Simplified Interface: Provide a more intuitive API that allows developers to use simple SQL statements or SQL-like syntax for querying, instead of requiring them to learn a complex DSL.
	2.	Good Documentation and Examples: Offer comprehensive documentation and practical examples to help developers get started quickly, while also explaining the underlying logic and common patterns.

These improvements could lower the entry barrier and make Rust more approachable for web development, ultimately attracting more developers to the ecosystem.

**Reply by bsodmike** [reply]
Interesting 🤔

> especially in a language like Rust, where the type system and borrow checker already have a certain learning curve

I find this as a rather curious comment, but for me personally, this is why I like Rust; the type-system is something we rely on for type-driven design, and there are ways around the borrowchk if needed (unsafe etc).  Anyways, it'll be interesting to see what other say here regarding your suggestion.

I'm just happy to see an ORM, especially supported by the Tokio working-group.

**Reply by cnwzhu** [reply]
Don’t you feel uncomfortable writing a bunch of Rust code for this query? Originally, similar functionality could be achieved with simple SQL, but now we have to implement a series of interfaces. Just because it’s complex doesn’t mean I like this style.

**Reply by cnwzhu** [reply]
https://github.com/bsodmike/toasty/blob/main/examples/hello-toasty/src/db/todo.rs

**Reply by bsodmike** [reply]
> https://github.com/bsodmike/toasty/blob/main/examples/hello-toasty/src/db/todo.rs

Nothing really jumps out at me, other than lifetime annotations. If you understand life times, and the other basics of Rust, it isn't that daunting.  One thing, the example you linked can be cleaned up quite a bit by adding the usual `use super::user::User` etc. declarations.

Beyond that, it looks pretty standard.  Anyone else have thoughts on this?

**Comment by xiaowan** [comment]
I also really like mybatis, simple and straightforward

**Comment by Reinhard-Berger** [comment]
I like the design of Doctrine ore eloquent in PHP (the problem i see is that php works with reflections and proxy objects in Doctrine. Which seems to me ver hard to reproduce in rust).


but i think we dont should copy a other style. Instead pick some good things and go the rust way

**Comment by maifuwa** [comment]
Maybe Jimmer of kotlin more comfortable
